thefrontside/effection
# ❌ Try out NX

It's no good. I'm going to switch to Turborepo.

<details>
<summary>Prev PR write up for NX</summary>
<br>
## Motivation

I was looking into configuring [NX](https://nx.dev/) for the `backstage` monorepo to speed up their build process but I wanted to try it out on a small monorepo first and thought it would be helpful for effection.

## Approach

NX is packed full of features (that I'm in the process of exploring) but this PR is to just introduce the very basic configuration which still offers huge benefits:

- `Computation caching`: If I run tests, and then run it again without making any changes to the code, NX knows the output will be the same and therefore won't force you to wait for the commands to run from scratch
  - I ran `test` and `lint` locally with and without NX to compare the difference in speed:
      | Command | First Run | Second Run |
      | -- | -- | -- |
      | `yarn workspaces run test` | 107 seconds | 106 seconds |
      | `nx run-many --target=test --all` | 32 seconds | < 1 second |
      | `yarn workspaces run lint` | 20 seconds | 20 seconds |
      | `nx run-many --target=lint --all` | 8 seconds | < 1 second |
- `Affected`: If the chart above isn't impressive enough, NX can also calculate which packages have been affected by the changes of your branch
  - If I run `nx affected --target=test` from my branch, it runs tests in every package because I've made changes to the root of the monorepo, but if you would like to see how `affected` works:
    1. pull the `mk/nx` branch
    2. run `yarn && yarn prepack`
    3. update the `defaultBase` property in `./nx.json` from `v2` to `mk/nx`
    4. run `yarn test` to see that it doesn't run any tests because there are no new changes
    5. throw in a `console.log` somewhere in `./packages/websocket-server/`
    6. run `yarn test` again and now it should have run tests for `inspect`, `websocket-server` and `inspect-server`
  - `affected` and `run-many` both run all tasks in parallel (3 by default but can be adjusted)
- ⚠️ I would've liked to replace the `prepack` script too but couldn't because of the issue with circular dependencies

## What's Next?

- [x] Update CircleCI config to utilize NX
- [x] Add NX Cloud - this allows you to share caching with other developers so if in CI or another developer runs a cacheable command, the cache gets uploaded to [nrwl](https://github.com/nrwl) and gets used if another developer runs that same command on another machine
  - [ ] Create an NX Cloud account to generate a new token
- [ ] Update required status checks for this repo
</details>
